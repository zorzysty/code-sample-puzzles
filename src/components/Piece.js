import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import imgSrc from '../assets/picture.jpg';
import { DIMENSIONS } from '../const';


const spec = {
  beginDrag: (props) => {
    props.handleDrag();
    return props.piece;
  },
  endDrag: (props, monitor) => {
    // return piece to original position when dropped outside of target
    if (!monitor.didDrop()) return;

    // pass piece to handleDrop method if placed correctly and false otherwise
    props.handleDrop(monitor.getDropResult().success && props.piece);
  },
};

function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  };
}

class Piece extends PureComponent {
  render() {
    const {
      isDragging, // eslint-disable-line react/prop-types
      connectDragSource, // eslint-disable-line react/prop-types
      style,
      piece,
    } = this.props;

    // hide piece in drawer while it's being dragged
    const opacity = isDragging ? 0 : 1;
    // prevent overlaying of images while dragging
    const zIndex = isDragging ? 0 : 10;
    // set proper piece image using CSS sprites
    const left = -(piece.id % DIMENSIONS.columns) * DIMENSIONS.pieceSize;
    const top = -Math.floor(piece.id / DIMENSIONS.rows) * DIMENSIONS.pieceSize;

    return connectDragSource(
      <div
        className="Piece"
        style={{
          background: `url(${imgSrc})`,
          backgroundPosition: `${left}px ${top}px`,
          backgroundSize: `${DIMENSIONS.columns * 100}%`,
          opacity,
          zIndex,
          ...style,
        }}
      />,
    );
  }
}

export default DragSource('piece', spec, collect)(Piece);

Piece.propTypes = {
  piece: PropTypes.shape({
    id: PropTypes.number.isRequired,
  }).isRequired,
  handleDrop: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  handleDrag: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  style: PropTypes.objectOf(PropTypes.string),
};

Piece.defaultProps = {
  handleDrop: () => {},
  handleDrag: () => {},
  style: {},
};
