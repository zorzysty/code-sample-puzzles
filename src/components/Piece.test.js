import React from 'react';
import { mount, shallow } from 'enzyme';
import Piece from './Piece';

describe('Piece', () => {
  it('should render correctly based on passed props', () => {
    const propsArray = [
      {
        key: 1,
        piece: { id: 1 },
        style: {
          top: '30px',
          left: '50px',
        },
      },
      {
        key: 0,
        piece: { id: 0 },
        style: {
          top: '70px',
          left: '10px',
        },
      },
    ];

    propsArray.forEach((props) => {
      const component = shallow(
        <Piece
          handleDrop={() => {}}
          handleDrag={() => {}}
          {...props}
        />,
      );

      expect(component)
        .toMatchSnapshot();

      component.unmount();
    });
  });

  it('should call functions passed in props', () => {
    const handleDragMock = jest.fn();
    const handleDropMock = jest.fn();

    const piece = mount(<Piece
      key={1}
      piece={{ id: 1 }}
      handleDrop={handleDropMock}
      handleDrag={handleDragMock}
    />);

    piece.prop('handleDrag')();
    expect(handleDragMock)
      .toHaveBeenCalledTimes(1);

    piece.prop('handleDrop')();
    expect(handleDropMock)
      .toHaveBeenCalledTimes(1);

    piece.unmount();
  });
});
