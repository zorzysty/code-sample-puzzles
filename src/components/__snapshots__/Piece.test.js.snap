// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piece should render correctly based on passed props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DragSource(Piece)
    handleDrag={[Function]}
    handleDrop={[Function]}
    piece={
        Object {
            "id": 1,
          }
    }
    style={
        Object {
            "left": "50px",
            "top": "30px",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "dragDropManager": undefined,
        },
        "_currentValue2": Object {
          "dragDropManager": undefined,
        },
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "dragDropManager": undefined,
          },
          "_currentValue2": Object {
            "dragDropManager": undefined,
          },
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Piece should render correctly based on passed props 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DragSource(Piece)
    handleDrag={[Function]}
    handleDrop={[Function]}
    piece={
        Object {
            "id": 0,
          }
    }
    style={
        Object {
            "left": "10px",
            "top": "70px",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "dragDropManager": undefined,
        },
        "_currentValue2": Object {
          "dragDropManager": undefined,
        },
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "dragDropManager": undefined,
          },
          "_currentValue2": Object {
            "dragDropManager": undefined,
          },
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
